[AMDGPU] Implement IR expansion for frem instruction

frederik-h opened the pull request:
This patch implements a correctly rounded expansion of the frem instruction in LLVM IR. This is useful for target architectures for which such an expansion is too involved to be implement in ISel Lowering. The expansion is based on the code from the AMD device libs and has been tested successfully against the OpenCL conformance tests on amdgpu. The expansion is implemented in the preexisting "expand-fp" pass. It replaces the expansion of "frem" in ISel for the amdgpu target; it is enabled for targets which do not directly support "frem" and for which no matching "fmod" LibCall is available.

frederik-h committed:
Implement IR expansion for frem instruction

This patch implements a correctly rounded expansion of the frem
instruction in LLVM IR. This is useful for target architectures where
such an expansion is too involved to be implement on ISel
Lowering. The expansion is based on the code from the AMD device libs
and has been tested successfully against the OpenCL conformance tests
on AMDGPU. The expansion is implemented in the preexisting
"expand-large-fp-convert" pass. It is enabled by a new
"shouldExpandFRemInIR" function in TargetLowering.

frederik-h committed:
Adjust some comments, remove include

frederik-h committed:
clang-format changes

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -5671,6 +5671,10 @@ class TargetLowering : public TargetLoweringBase {
 
+  /// Indicates whether the FRem instruction should be expanded before
+  /// ISel in the LLVM IR.
+  virtual bool shouldExpandFRemInIR() const { return false; };
```
Don't introduce a new hook for this. You can check if the operation is going to expand, and then check that there is no call in RuntimeLibcalls (this is essentially how PreISelIntrinsicLowering handles memcpy and co)

frederik-h replied:
I see that I can check whether the operation is going to expand using `TargetLowering::isOperationExpand`.  But there is no `RTLIB::Libcall` enum value for the `frem` instruction in `llvm/include/llvm/IR/RuntimeLibcalls.def"` which is what we are trying to expand here; of course, we could include the `fmod` variants here. So, for `frem`, I would just check for  `!TargetLowering::isOperationExpand`, right?

frederik-h replied:
I got it now. I needed to map `frem` to the right `fmod` LibCall.

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -28,6 +29,9 @@
 #include "llvm/Target/TargetMachine.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+#define DEBUG_TYPE "expand-large-fp-convert"
```
Should rename the pass to something else preferably in a pre-commit

frederik-h replied:
Done in PR #131128.

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+        One(ConstantInt::get(ExTy, 1)), Signbit(Signbit) {};
+
+  Value *createLdexp(Value *Base, Value *Exp, const Twine &Name) const {
+    return B.CreateIntrinsic(Intrinsic::ldexp, {ComputeFpTy, B.getInt32Ty()},
+                             {Base, Exp}, {}, Name);
```
Should move this into IRBuilder

frederik-h replied:
This one (i.e. `IRBuilder::CreateLdexp`) exists in `IRBuilder`. I'll change the code here to use it.

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+  }
+
+  Value *createRcp(Value *V, const Twine &Name) const {
+    return B.CreateFDiv(ConstantFP::get(ComputeFpTy, 1.0), V, Name);
```
Can / should we do something to make this a fast rcp

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+    //   ax = clt ? axp : ax;
+    Value *Q = B.CreateUnaryIntrinsic(Intrinsic::rint, B.CreateFMul(Ax, Ayinv),
+                                      {}, "q");
+    Value *AxUpdate = B.CreateIntrinsic(Intrinsic::fma, {ComputeFpTy},
+                                        {B.CreateFNeg(Q), Ay, Ax}, {}, "ax");
```
I would hope IRBuilder has a CreateFMA

frederik-h replied:
There wasn't one, but see #131112.

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+    // TODO Handle NaN and infinity fast math flags separately here?
+    Value *Nan = ConstantFP::getQNaN(FremTy);
+
+    Ret = B.CreateSelect(B.createIsFPClass(Y, FPClassTest::fcZero), Nan, Ret);
+    Value *C = B.CreateLogicalAnd(
+        B.CreateNot(B.createIsFPClass(Y, FPClassTest::fcNan)),
+        B.createIsFPClass(X, FPClassTest::fcFinite));
+    Ret = B.CreateSelect(C, Ret, Nan);
```
This expression ought to be cleaned up. We canonically prefer using fcmp over is.fpclass, and not of is.fpclass should fold into the fpclass mask. I think InstCombine handles all of this now, you can run the IR output through it to see a simpler form

frederik-h replied:
I did this and I also resolved my "TODO Handle NaN and infinity fast math flags separately here?" at the same time.

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+  PHINode *RetPhi = B.CreatePHI(FremTy, 2, "ret");
+  Value *Ret = RetPhi;
+
+  if (!FMF.noNaNs() || !FMF.noInfs())
```
TODO use isKnownNeverNaN

frederik-h replied:
Done

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+/// Returns true if \p I should not be expanded because
+/// it will be eliminated during ISel.
+static bool shouldSkipExpandFRem(BinaryOperator &I) {
+  // This condition should be sufficient for DAGCombiner::visitFREM to
+  // eliminate the instruction.
```
We can't rely on this for correctness. The pass should just fully process the frems

frederik-h replied:
I also consider this to be a temporary workaround (should have added at least a TODO ...). Without this, there is a regression in this test case: `https://github.com/llvm/llvm-project/blob/139add531a5533a7591527a3e6cee7dca0f795c3/llvm/test/CodeGen/AMDGPU/dagcombine-select.ll#L302`. I assume we would like to keep folding constant frem uses in this way. Perhaps this folding can be replicated on the IR level?

frederik-h replied:
Should we accept missing the optimization for now and replicate some of the optimizations at the IR level later? There are other optimizations in the backend which we probably want to handle here, e.g. using a simple formula without any loops if the denumerator is a power of 2.

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+  if (ReturnTy->isFloatingPointTy())
+    Ret = Expander->buildFRem(I.getOperand(0), I.getOperand(1));
+  else {
+    auto VecTy = cast<FixedVectorType>(ReturnTy);
```
```suggestion
    auto *VecTy = cast<FixedVectorType>(ReturnTy);
```

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -37,6 +41,377 @@ static cl::opt<unsigned>
+
+  I.replaceAllUsesWith(Ret);
+  Ret->takeName(&I);
+  I.removeFromParent();
+  I.dropAllReferences();
```
```suggestion
  I.eraseFromParent();
```

arsenm commented on llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:
```diff
@@ -654,8 +1035,10 @@ static bool runImpl(Function &F, const TargetLowering &TLI) {
-    if (I->getOpcode() == Instruction::FPToUI ||
-        I->getOpcode() == Instruction::FPToSI) {
+    if (I->getOpcode() == Instruction::FRem)
+      expandFRem(llvm::cast<BinaryOperator>(*I));
```
```suggestion
      expandFRem(cast<BinaryOperator>(*I));
```

arsenm reviewed (commented):
Missing test changes. We probably need to replace whatever few frem tests already exist. Also need checking of the IR output of the expansion

frederik-h committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Rename ExpandLargeFpConvertPass to ExpandFpPass

This is meant as a preparation for PR #130988 "[AMDGPU] Implement IR expansion for frem instruction" which implements the expansion of another instruction in this pass. The more general name seems more appropriate given this change and quite reasonable even without it.

The renaming of the source files happens in the following commit for a
better diff.

* llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:

frederik-h committed:
Rename ExpandLargeFpConvert.{cpp,h}

frederik-h committed:
clang-format changes

Adjust line length

frederik-h committed:
Add IRBuilder::CreateFMA (#131112)

This commit adds a function for creating fma intrinsic calls to the IRBuilder.  If the "IsFPConstrained" flag of the builder is set,
the function creates a call to "experimental.constrained.fma" instead of "llvm.fma" .
To support the creation of the constrained intrinsic, a function "CreateConstrainedFPIntrinsic" is introduced.

frederik-h committed:
Review changes

Use IRBuilder functions for creating Ldexp, FMA intrinsic call

Use different condition for controlling frem expansion

Update test

Fixup after merge

NaN handling fixes

Update tests

arsenm committed:
AMDGPU: Add more freeze codegen tests

arsenm committed:
DAG: Fix promote of half freeze

Round out the AMDGPU codegen test to all the generations to cover
the illegal f16 targets.

arsenm committed:
Fix losing the actual freeze

frederik-h committed:
Adjust expand-fp pass DEBUG_TYPE

frederik-h committed:
Use AssumptionCache

... only if optimizations are enabled. Disable Value Tracking analysis
if optimizations are disabled.

frederik-h committed:
Adjust freeze.ll tst

frederik-h committed:
Remove exceptional handling to dagcombine

frederik-h committed:
Adjust test

frederik-h committed:
clang-format changes

frederik-h committed:
Adjust type

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into frem-expansion

frederik-h committed:
Update tests

frederik-h committed:
Adjust type to fix Windows compilation

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +48,352 @@ static cl::opt<unsigned>
+    Value *AxAsInt = B.CreateBitCast(AxFinal, IntTy, "ax_as_int");
+
+    Value *Ret =
+        B.CreateXor(B.CreateAnd(XAsInt, Signbit), AxAsInt, "Remainder");
```
Introduce a CreateSignBit function, based on clang's handling of __builtin_signbit this requires more work for some FP types

frederik-h committed:
review changes

frederik-h committed:
Adjust comments

Replace macros and functions from device-libs by shorter and more
compact suggestive names in comments.

frederik-h committed:
fixup! Adjust comments

frederik-h commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +48,352 @@ static cl::opt<unsigned>
+public:
+  static std::optional<FRemExpander> create(IRBuilder<> &B, Type *Ty) {
+    if (Ty->is16bitFPTy())
+      return FRemExpander{B, Ty, 11, 0x8000, B.getFloatTy(), B.getInt16Ty()};
```
For `float`, this is half the precision and it means that the main compute loop will iterate (at most) twice. For the 16 bit type, this says that we want to compute all bits in one iteration which is possible because we use the extended precision type for the computation. I changed this code to use the `fltSemantics`.

frederik-h commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +48,352 @@ static cl::opt<unsigned>
+    FPClassTest NotNan = FPClassTest::fcInf | FPClassTest::fcFinite;
+    Value *YNotNan = SQ && isKnownNeverNaN(Y, 0, *SQ)
+                         ? B.getTrue()
+                         : B.createIsFPClass(Y, NotNan);
```
I rearranged the code a bit to merge the NaN check into the finiteness check for y using the fcmp ord.

frederik-h committed:
clang-format changes

frederik-h committed:
Update llvm/lib/CodeGen/ExpandFp.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/lib/CodeGen/ExpandFp.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/lib/CodeGen/ExpandFp.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Update llvm/test/CodeGen/AMDGPU/dagcombine-select.ll

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Remove unused #include

frederik-h committed:
Handle skipFunction and optnone

frederik-h committed:
Change inappropriate use of "auto"

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +49,340 @@ static cl::opt<unsigned>
+  // loop body and the "final iteration".
+  Value *buildUpdateAx(Value *Ax, Value *Ay, Value *Ayinv) const {
+    // Build:
+    //   float q = rint(ax * ayinv);
+    //   ax = fma(-q, ay, ax);
```
We appear to be missing folds of fneg with rint. In this case this works out, you can't save encoding size if you pull the fneg away from the fma. But in general we should probably have `fneg (rint(x)) -> rint(fneg(x))`

frederik-h replied:
I suppose this is a follow-up issue.

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +49,340 @@ static cl::opt<unsigned>
+                                      {}, "q");
+    Value *AxUpdate = B.CreateFMA(B.CreateFNeg(Q), Ay, Ax, {}, "ax");
+    Value *Clt = B.CreateFCmp(CmpInst::FCMP_OLT, AxUpdate,
+                              ConstantFP::get(ComputeFpTy, 0.0), "clt");
```
```suggestion
                              ConstantFP::getZero(ComputeFpTy), "clt");
```

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +49,342 @@ static cl::opt<unsigned>
+    // One iteration of the loop is factored out.  The code shared by
+    // the loop and this "iteration" is denoted by UPDATE_AX.
+    B.SetInsertPoint(LoopBB);
+    auto *NbIv = B.CreatePHI(Nb->getType(), 2, "nb_iv");
```
no auto

arsenm reviewed (commented):
Have you verified instcombine doesn't do anything with the expanded output?

frederik-h committed:
Update llvm/lib/CodeGen/ExpandFp.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Add dagcombine-select test

frederik-h committed:
Use isHalfyTy instead of is16bitFPTy

Try to be more specificb, as bfloat16 is not supported.

frederik-h committed:
Add opt-level parameter to ExpandFp pass

frederik-h commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -696,13 +1104,19 @@ class ExpandFpLegacyPass : public FunctionPass {
-                                               : PreservedAnalyses::all();
+  auto &TLI = *STI->getTargetLowering();
+  AssumptionCache *AC = nullptr;
+  if (TM->getOptLevel() != CodeGenOptLevel::None)
```
Should we use `CodeGenOptLevel::None` with the new pass manager if the parameter is not provided or should the default be `TM->getOptLevel()`?

arsenm replied:
I'm not sure, this area is kind of a mess. We have mixed uses of both

frederik-h commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -2393,21 +2393,6 @@ SDValue AMDGPUTargetLowering::LowerSDIVREM(SDValue Op,
   return DAG.getMergeValues(Res, DL);
 }
 
-// (frem x, y) -> (fma (fneg (ftrunc (fdiv x, y))), y, x)
```
No, would this be acceptable with `afn`?

arsenm replied:
Probably, it's not particularly well defined but it starts breaking down at pretty large values

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into frem-expansion

frederik-h committed:
Update tests

frederik-h committed:
clang-format changes

frederik-h committed:
Merge remote-tracking branch 'upstream' into frem-expansion

frederik-h committed:
Adjust to isKnownNeverInfinity changes

frederik-h committed:
Merge remote-tracking branch 'upstream' into frem-expansion

frederik-h committed:
Move check for ScalableVectorType earlier

frederik-h committed:
Adjust tests

True16 tests had to be disabled because of issue:
[AMDGPU][True16] si-fix-sgpr-copies: invalid sgpr_lo16 copy destination #144561

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into frem-expansion

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into frem-expansion

frederik-h committed:
Introduce approximate FRem expansion

frederik-h commented:
@arsenm ping

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +48,358 @@ static cl::opt<unsigned>
+
+Value *FRemExpander::buildApproxFRem(Value *X, Value *Y) const {
+  IRBuilder<>::FastMathFlagGuard Guard(B);
+  // Propagating the approximate functions flag to the
+  // division leads to an unacceptable drop in precision
+  // on AMDGPU.
+  B.clearFastMathFlags();
```
I think this can be more specific, at least some of the flags can be preserved. I believe we can use one of the faster fdiv paths but I don't remember the details. Would require further investigation in a follow up

frederik-h replied:
I have added a TODO as a reminder.

arsenm commented on llvm/test/CodeGen/AMDGPU/dagcombine-select-frem.ll:
```diff
@@ -0,0 +1,14 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 -verify-machineinstrs < %s | FileCheck -enable-var-scope -check-prefix=GCN %s
```
```suggestion
; RUN: llc -mtriple=amdgcn -mcpu=gfx900 < %s | FileCheck -enable-var-scope -check-prefix=GCN %s
```

arsenm commented on llvm/test/CodeGen/AMDGPU/dagcombine-select-frem.ll:
```diff
@@ -0,0 +1,14 @@
+  %bo = frem float 5.0, %sel
+  store float %bo, ptr addrspace(1) %p, align 4
+  ret void
+}
```
```suggestion
}

```

arsenm commented on llvm/test/CodeGen/AMDGPU/dagcombine-select-frem.ll:
```diff
@@ -0,0 +1,14 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 -verify-machineinstrs < %s | FileCheck -enable-var-scope -check-prefix=GCN %s
+; XFAIL: *
```
No xfailed test, especially one with checks

frederik-h replied:
This was extracted from dagcombine-select.ll. I deleted this file and added moved the TODO to the original test case in the former file.

arsenm commented on llvm/test/Transforms/ExpandFp/AMDGPU/frem.ll:
```diff
@@ -0,0 +1,1375 @@
+}
+
+attributes #0 = { nounwind "unsafe-fp-math"="false" "denormal-fp-math-f32"="preserve-sign,preserve-sign" }
+attributes #1 = { nounwind "unsafe-fp-math"="true" "denormal-fp-math-f32"="preserve-sign,preserve-sign" }
```
Don't introduce new uses of "unsafe-fp-math", only the flag matters now

arsenm replied:
bump, this is being removed

frederik-h replied:
> bump, this is being removed

I am not sure if I understand what you mean? Was this a reminder to remove the attributes before merging? In any case, that's what I did.

arsenm replied:
yes, there was an update but they were still there. I mean the attribute support is being dropped from codegen

arsenm reviewed (approved):
LGTM with some test cleanups

frederik-h committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into frem-expansion

frederik-h committed:
Review changes

* Add TODO concerning propagation of fmf
* Remove attributes from ExpandFp test
* Remove xfailed dagcombine-select-frem.ll and add TODO to corresponding
test in dagcombine-select.ll.

frederik-h committed:
Remove extra ; to fix CI build error

The CI build has started failing because of this ; which has been
there for quite some time:

  /home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/CodeGen/ExpandFp.cpp:1134:36: error: extra ';' outside of a function is incompatible with C++98 [-Werror,-Wc++98-compat-extra-sem
  1134 |     : TM(TM), OptLevel(OptLevel) {};

Remove it.

mikaelholmen commented:
Hi @frederik-h 

The following starts crashing with this patch:
```
llc bbi-110248.ll
```
It fails like
```
llc: ../lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:5858: SDValue llvm::DAGTypeLegalizer::ExpandIntOp_XINT_TO_FP(SDNode *): Assertion `LC != RTLIB::UNKNOWN_LIBCALL && "Don't know how to expand this XINT_TO_FP!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build-all/bin/llc -mtriple x86_64-unknown-linux-gnu bbi-110248.ll
1.	Running pass 'Function Pass Manager' on module 'bbi-110248.ll'.
2.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@main'
 #0 0x0000556515aa33a6 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/llc+0x7bd33a6)
 #1 0x0000556515aa0af5 llvm::sys::RunSignalHandlers() (build-all/bin/llc+0x7bd0af5)
 #2 0x0000556515aa41b9 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007f42a15a4990 __restore_rt (/lib64/libpthread.so.0+0x12990)
 #4 0x00007f429ef4452f raise (/lib64/libc.so.6+0x4e52f)
 #5 0x00007f429ef17e65 abort (/lib64/libc.so.6+0x21e65)
 #6 0x00007f429ef17d39 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d39)
 #7 0x00007f429ef3ce86 (/lib64/libc.so.6+0x46e86)
 #8 0x00005565158c03f5 llvm::DAGTypeLegalizer::ExpandIntOp_XINT_TO_FP(llvm::SDNode*) LegalizeIntegerTypes.cpp:0:0
 #9 0x00005565158bf15f llvm::DAGTypeLegalizer::ExpandIntegerOperand(llvm::SDNode*, unsigned int) LegalizeIntegerTypes.cpp:0:0
#10 0x0000556515884c8d llvm::DAGTypeLegalizer::run() LegalizeTypes.cpp:0:0
#11 0x000055651588a449 llvm::SelectionDAG::LegalizeTypes() (build-all/bin/llc+0x79ba449)
#12 0x0000556515870350 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (build-all/bin/llc+0x79a0350)
#13 0x000055651586ed9e llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (build-all/bin/llc+0x799ed9e)
#14 0x000055651586c371 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (build-all/bin/llc+0x799c371)
#15 0x0000556515869d39 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) (build-all/bin/llc+0x7999d39)
#16 0x0000556514ab5ee7 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (build-all/bin/llc+0x6be5ee7)
#17 0x000055651500ee4c llvm::FPPassManager::runOnFunction(llvm::Function&) (build-all/bin/llc+0x713ee4c)
#18 0x00005565150170f2 llvm::FPPassManager::runOnModule(llvm::Module&) (build-all/bin/llc+0x71470f2)
#19 0x000055651500f918 llvm::legacy::PassManagerImpl::run(llvm::Module&) (build-all/bin/llc+0x713f918)
#20 0x00005565129cef87 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#21 0x00005565129cc500 main (build-all/bin/llc+0x4afc500)
#22 0x00007f429ef307e5 __libc_start_main (/lib64/libc.so.6+0x3a7e5)
#23 0x00005565129cb96e _start (build-all/bin/llc+0x4afb96e)
Abort (core dumped)
```
I don't know what x86 thinks about i224 as input arguments but this didn't crash before.
We see this all over the place in downstream fuzzy testing.
[bbi-110248.ll.gz](https://github.com/user-attachments/files/22136413/bbi-110248.ll.gz)

bjope commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -720,6 +719,13 @@ FUNCTION_PASS_WITH_PARAMS(
+    [TM = TM](CodeGenOptLevel OL) {
+      return ExpandFpPass(TM, OL);
+    },
+    parseExpandFpOptions, "opt-level")
```
I think this could say "opt-level=N" to indicate that the syntax is like that, and not that opt-level should be replaced by a value. Similar to "guard=N" for bounds-checking.

bjope commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -720,6 +719,13 @@ FUNCTION_PASS_WITH_PARAMS(
     },
     parseBoundsCheckingOptions,
     "trap;rt;rt-abort;min-rt;min-rt-abort;merge;guard=N")
+    FUNCTION_PASS_WITH_PARAMS(
```
Wrong indents.

bjope commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -672,36 +1095,68 @@ static bool runImpl(Function &F, const TargetLowering &TLI) {
-    return runImpl(F, *TLI);
+    AssumptionCache *AC = nullptr;
+
+    if (OptLevel != CodeGenOptLevel::None || F.hasOptNone())
```
Isn't this a bit weird? (If OptLevel is O1/O2/O3 or if function attributes say O0.)

If F.hasOptNone() is true then normally skipFunction would return true so we wouldn't even get here.

arsenm replied:
It is invalid to use skipFunction in mandatory lowering passes

mikaelholmen commented:
> Hi @frederik-h
> 
> The following starts crashing with this patch:
> 
> ```
> llc -mtriple x86_64-unknown-linux-gnu bbi-110248.ll
> ```
> 
> It fails like
> 
> ```
> llc: ../lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:5858: SDValue llvm::DAGTypeLegalizer::ExpandIntOp_XINT_TO_FP(SDNode *): Assertion `LC != RTLIB::UNKNOWN_LIBCALL && "Don't know how to expand this XINT_TO_FP!"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> Stack dump:
> 0.	Program arguments: build-all/bin/llc -mtriple x86_64-unknown-linux-gnu bbi-110248.ll
> 1.	Running pass 'Function Pass Manager' on module 'bbi-110248.ll'.
> 2.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@main'
>  #0 0x0000556515aa33a6 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/llc+0x7bd33a6)
>  #1 0x0000556515aa0af5 llvm::sys::RunSignalHandlers() (build-all/bin/llc+0x7bd0af5)
>  #2 0x0000556515aa41b9 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
>  #3 0x00007f42a15a4990 __restore_rt (/lib64/libpthread.so.0+0x12990)
>  #4 0x00007f429ef4452f raise (/lib64/libc.so.6+0x4e52f)
>  #5 0x00007f429ef17e65 abort (/lib64/libc.so.6+0x21e65)
>  #6 0x00007f429ef17d39 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d39)
>  #7 0x00007f429ef3ce86 (/lib64/libc.so.6+0x46e86)
>  #8 0x00005565158c03f5 llvm::DAGTypeLegalizer::ExpandIntOp_XINT_TO_FP(llvm::SDNode*) LegalizeIntegerTypes.cpp:0:0
>  #9 0x00005565158bf15f llvm::DAGTypeLegalizer::ExpandIntegerOperand(llvm::SDNode*, unsigned int) LegalizeIntegerTypes.cpp:0:0
> #10 0x0000556515884c8d llvm::DAGTypeLegalizer::run() LegalizeTypes.cpp:0:0
> #11 0x000055651588a449 llvm::SelectionDAG::LegalizeTypes() (build-all/bin/llc+0x79ba449)
> #12 0x0000556515870350 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (build-all/bin/llc+0x79a0350)
> #13 0x000055651586ed9e llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (build-all/bin/llc+0x799ed9e)
> #14 0x000055651586c371 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (build-all/bin/llc+0x799c371)
> #15 0x0000556515869d39 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) (build-all/bin/llc+0x7999d39)
> #16 0x0000556514ab5ee7 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (build-all/bin/llc+0x6be5ee7)
> #17 0x000055651500ee4c llvm::FPPassManager::runOnFunction(llvm::Function&) (build-all/bin/llc+0x713ee4c)
> #18 0x00005565150170f2 llvm::FPPassManager::runOnModule(llvm::Module&) (build-all/bin/llc+0x71470f2)
> #19 0x000055651500f918 llvm::legacy::PassManagerImpl::run(llvm::Module&) (build-all/bin/llc+0x713f918)
> #20 0x00005565129cef87 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
> #21 0x00005565129cc500 main (build-all/bin/llc+0x4afc500)
> #22 0x00007f429ef307e5 __libc_start_main (/lib64/libc.so.6+0x3a7e5)
> #23 0x00005565129cb96e _start (build-all/bin/llc+0x4afb96e)
> Abort (core dumped)
> ```
> 
> I don't know what x86 thinks about i224 as input arguments but this didn't crash before. We see this all over the place in downstream fuzzy testing. [bbi-110248.ll.gz](https://github.com/user-attachments/files/22136413/bbi-110248.ll.gz)

When running "expand-fp" it just says
```
Skipping pass 'Expand fp' on function main
```
with this patch, but before it expanded it to a lot of stuff.

piotrAMD commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -37,6 +48,359 @@ static cl::opt<unsigned>
+
+  Value *Ret;
+  if (ReturnTy->isFloatingPointTy())
+    Ret = FMF.approxFunc()
```
@frederik-h @arsenm We are seeing some cts failures from this patch. We're not setting _afn_ on a double frem (`%24 = frem arcp double %22, %23`) and the test expects flush to zero behaviour ("preserve-sign" set in the func attr). After the patch, instead of 0 we get a denorm because non-fast version is selected.
Just curious - should this condition be extended to somehow include DenormalMode?

arsenm replied:
First "preserve-sign" indicates the default rounding mode; it explicitly does not mandate flushing.
Second, are you really depending on the behavior of flush with double? I expect AMDGPU to only ever use the flushing modes with float

piotrAMD replied:
Thanks, so perhaps it just worked by accident, and this patch just revealed the issue, because it changed the behavior.
There are Vulkan tests that require FTZ behavior for 64-bit (`OpExecutionMode %main DenormFlushToZero 64`).

But I still wonder if this patch handles flushing mods with float correctly.
